crucible-llvm/src/Lang/Crucible/LLVM/MemModel/Strings.hs

```diff
@@ -3,27 +3,261 @@
 import qualified Lang.Crucible.LLVM.MemModel as Mem
+import qualified Lang.Crucible.LLVM.MemModel.Partial as Partial
+import qualified What4.Expr.Builder as WEB
 import qualified What4.Interface as WI
+import qualified What4.Protocol.Online as WPO
+import qualified What4.SatResult as WS
+
+-- | Whether to stop or keep going
+--
+-- Like Rust's @std::ops::ControlFlow@.
+data ControlFlow a b
+  = Continue a
+  | Break b
+  deriving Functor
+
+instance Bifunctor ControlFlow where
+  bimap l r =
+    \case
+      Continue x -> Continue (l x)
+      Break x -> Break (r x)
+
+-- | Compute a result from a symbolic byte, and check if the load should
+-- continue to the next byte.
+--
+-- Used to:
+--
+-- * Check if the byte is concretely a null terminator
+--   ('concretelyNullTerminatedString')
+-- * Check if a byte is known by a solver to be a null terminator ('nullTerminatedString')
+-- * Check if we have surpassed a length limit ('withMaxChars')
+--
+-- Note that it is relatively common for @a@ to be a function @[b] -> [b]@. This
+-- is used to build up a snoc-list.
+newtype ByteChecker m sym bak a b
+  = ByteChecker { runByteChecker :: bak -> a -> Mem.LLVMPtr sym 8 -> m (ControlFlow a b) }
+
+-- | Load a sequence of bytes, one at a time.
+--
+-- Not exported.
+loadBytes ::
+  forall m a b sym bak wptr.
+  ( LCB.IsSymBackend sym bak
+  , Mem.HasPtrWidth wptr
+  , Mem.HasLLVMAnn sym
+  , ?memOpts :: Mem.MemOptions
+  , GHC.HasCallStack
+  , MonadIO m
+  ) =>
+  bak ->
+  Mem.MemImpl sym ->
+  -- | Initial accumulator
+  a ->
+  Mem.LLVMPtr sym wptr ->
+  ByteChecker m sym bak a b ->
+  m b
+loadBytes bak mem = go
+ where
+  sym = LCB.backendGetSym bak
+  go ::
+    a ->
+    Mem.LLVMPtr sym wptr ->
+    ByteChecker m sym bak a b ->
+    m b
+  go acc ptr f = do
+    let i1 = Mem.bitvectorType 1
+    let p8 = Mem.LLVMPointerRepr (DPN.knownNat @8)
+    byte <- liftIO (Mem.doLoad bak mem ptr i1 p8 CLD.noAlignment)
+    runByteChecker f bak acc byte >>=
+      \case
+        Break result -> pure result
+        Continue acc' -> do
+          ptr' <- liftIO (Mem.doPtrAddOffset bak mem ptr =<< WI.bvOne sym Mem.PtrWidth)
+          go acc' ptr' f
+
+ptrToBv8 ::
+  LCB.IsSymBackend sym bak =>
+  bak ->
+  Mem.LLVMPtr sym 8 ->
+  IO (WI.SymBV sym 8)
+ptrToBv8 bak bytePtr = do
+  let err = LCS.AssertFailureSimError "Found pointer instead of byte when loading string" ""
+  Partial.ptrToBv bak err bytePtr
+
+-- Currently unused, but analogous with 'Lang.Crucible.LLVM.MemModel.loadString'
+_fullyConcreteNullTerminatedString ::
+  GHC.HasCallStack =>
+  LCB.IsSymBackend sym bak =>
+  ByteChecker IO sym bak ([Word8] -> [Word8]) [Word8]
+_fullyConcreteNullTerminatedString =
+  ByteChecker $ \bak acc bytePtr -> do
+    byte <- ptrToBv8 bak bytePtr
+    case BV.asUnsigned <$> WI.asBV byte of
+      Just 0 -> pure (Break (acc []))
+      Just c -> do
+        let c' = toEnum @Word8 (fromInteger c)
+        pure (Continue (acc . (c':)))
```

Comment:
Similar comments here (and in related functions) about `acc . (c':)` being a bit too pointfree for my liking.